feat: midas ark

halaprix opened the pull request:
## Description
This PR implements the `MidasArk`, a new Ark integration for the Summer Earn Protocol that enables yield generation through Midas Protocol vaults.

The implementation follows the `ArkWithWithdrawalRequest` pattern, handling the specific interactions required for Midas's dual-vault architecture (Issuance Vault for deposits, Redemption Vault for withdrawals) and their mToken share system.

## Changes
- Added `MidasArk.sol`: The core contract implementing `IArk` and inheriting `ArkWithWithdrawalRequest`.
- Implemented `board()` logic to deposit assets into the Midas Issuance Vault (`depositInstant`).
- Implemented `requestWithdrawal()` to handle asynchronous redemptions via the Midas Redemption Vault.
- Implemented `withdrawUsingSwap()` to support instant withdrawals via the Redemption Vault's instant redemption mechanism (treated as a swap).
- Integrated `IMidasOracle` for accurate share-to-asset valuation.
- Added comprehensive fork tests in `MidasArkUSDC.fork.t.sol` covering boarding, withdrawals, and value accounting.

## Benefits
1. **New Yield Source**: Enables users to earn yield from Midas Protocol strategies (e.g., mTBILL/mBASIS) through the Summer Earn interface.
2. **Flexible Exit Options**: Supports both standard asynchronous redemptions (primary flow) and instant exits via the "withdraw using swap" mechanism.
3. **Accurate Valuation**: Uses Midas's official oracle for precise TVL and share price calculations, ensuring fair value during deposits and withdrawals.

## Testing
- **Fork Tests**: Verified against mainnet using `MidasArkUSDC.fork.t.sol`.
  - `test_Board_Midas_fork`: Verifies deposits into the Issuance Vault.
  - `test_RequestPartialRedeem_Midas_fork`: Verifies queuing of withdrawal requests.
  - `test_WithdrawUsingSwap_Midas`: Verifies instant redemption capabilities.
  - `test_TotalAssets_UsesOracle`: Ensures asset reporting matches oracle rates.

To run tests:
forge test --match-contract MidasArkTestFork

halaprix committed:
feat: introduce MidasArk contract and related interfaces

- Implemented the MidasArk contract for managing token supply and yield generation through Midas vaults.
- Added interfaces for DepositVault, RedemptionVault, MidasOracle, and MidasWithdrawalManager to facilitate interactions with the Midas protocol.
- Created a test suite for MidasArk to validate functionality including asset boarding, withdrawal requests, and total asset calculations using oracle prices.
- Established error handling for invalid vault addresses and ensured proper initialization of state variables.

halaprix committed:
feat: enhance MidasArk contract with improved withdrawal logic and constants

- Added RequestStatus import to manage redemption request states.
- Updated assetsInWithdrawalQueue function to check for pending requests before returning assets.
- Refactored requestWithdrawal function to prevent multiple withdrawal requests.
- Introduced MIDAS_ROUNDING_OFFSET constant for slippage calculations.
- Added internal functions for validating board and disembark data, ensuring no additional parameters are required.

halaprix committed:
fix: remove unnecessary console log in MidasArk test and update pendle-core subproject status

- Removed the console log statement for shares calculation in MidasArk test to clean up output.
- Updated pendle-core subproject commit to indicate a dirty state, reflecting local changes.

halaprix committed:
refactor: reorganize MidasArk contract structure and enhance functionality

- Updated import statements for clarity and added necessary interfaces.
- Refactored state variables into distinct sections for better organization.
- Improved asset calculation methods by introducing internal functions for shares to assets conversion.
- Enhanced withdrawal logic to streamline request handling and ensure accurate asset retrieval.
- Removed redundant code and comments to improve readability and maintainability.

halaprix committed:
refactor: clean up MidasArk contract by removing redundant comments

- Removed unnecessary comment regarding total shares calculation in MidasArk contract to enhance code clarity and maintainability.
- Updated pendle-core subproject status to reflect a dirty state, indicating local modifications.

actions-user committed:
Update Codecov badge to current branch

robercano commented on packages/core-contracts/src/contracts/arks/MidasArk.sol:
```diff
@@ -0,0 +1,342 @@
+
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+error InvalidIssuanceVault();
```
Chain Security audit pointed out that errors should be defined in a separate file. It is also my preference. What do you think about it?

robercano commented on packages/core-contracts/src/contracts/arks/MidasArk.sol:
```diff
@@ -0,0 +1,342 @@
+        // Calculate conversion factor
+        uint8 mTokenDecimals = mToken.decimals();
+        uint8 assetDecimals = IERC20Metadata(address(config.asset)).decimals();
+        if (mTokenDecimals < assetDecimals) revert InvalidMTokenDecimals();
```
Why is this? We could have the converting factor accepting mToken having less decimals than the asset.

robercano commented on packages/core-contracts/src/contracts/arks/MidasArk.sol:
```diff
@@ -0,0 +1,342 @@
+        return IERC20(config.asset).balanceOf(address(this));
+    }
+
+    function _board(uint256 amount, bytes calldata) internal override {
```
This should have a 

```
/**
 * @inherited ArkWithWithdrawalRequest
 */ 
```

robercano commented on packages/core-contracts/src/contracts/arks/MidasArk.sol:
```diff
@@ -0,0 +1,342 @@
+     * @dev The data can be empty as we don't use additional parameters
+     * @param /// data Additional data to validate
+     */
+    function _validateBoardData(bytes calldata) internal pure override {}
```
I'd love to se a `// no-op` here

robercano commented on packages/core-contracts/src/interfaces/midas/IDepositVault.sol:
```diff
@@ -0,0 +1,284 @@
+     */
+    function setMaxSupplyCap(uint256 newValue) external;
+
+    function tokensConfig(
```
Can we document these functions, even if it is to say that there is nothing to document? It stands out when reading the interface

robercano commented on packages/core-contracts/src/interfaces/midas/IMidasOracle.sol:
```diff
@@ -0,0 +1,37 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+interface IMidasOracle {
```
Same here?

robercano reviewed (changes requested):
Looks great! I've gone very specific to iterate on our best practices and have a discussion about what we want to do in the code. Also, what's the plan for the test TODOs?

halaprix committed:
Merge branch 'main' into kk/midas-ark

halaprix committed:
feat: introduce HyperBeatCoreArk and MidasArk contracts

- Added HyperBeatCoreArk contract to facilitate integration with the HyperBeatCore protocol.
- Implemented MidasArk contract for managing Midas protocol interactions.
- Created interfaces for HyperBeat components including IHyperBeatDepositor, IHyperBeatOracle, IHyperBeatPricer, and IHyperBeatVaultToken.
- Updated deployment scripts and configuration files to support new ark types and their parameters.
- Enhanced test coverage for the new contracts and their functionalities.

halaprix committed:
refactor: implement custom error handling in HyperBeatCoreArk and MidasArk contracts

- Introduced custom error interfaces IHyperBeatCoreArkErrors and IMidasArkErrors for better error management.
- Replaced existing revert statements with specific error calls in HyperBeatCoreArk and MidasArk constructors.
- Enhanced clarity and maintainability of error handling in both contracts.
- Updated documentation for withdrawal functions to clarify special withdrawal conditions.

halaprix committed:
docs: enhance comments for decimal checks in HyperBeatCoreArk and MidasArk contracts

- Added clarifying comments regarding the handling of token decimals in the conversion factor calculations.
- Specified that vault tokens and mTokens are expected to have 18 decimals, and noted the rarity of assets with more than 18 decimals in the context of these contracts.

halaprix commented on packages/core-contracts/src/contracts/arks/MidasArk.sol:
```diff
@@ -0,0 +1,342 @@
+        uint256 shares,
+        uint256 price
+    ) internal view returns (uint256) {
+        return (shares * price) / (TO_M_TOKEN_DECIMALS * Constants.WAD);
```
This could also be part of another PR, as PriceUtils was not maintained and uses PRB math 2 major versions behind the repository's version.

halaprix committed:
feat: add MidasArk to BummerLazyVault_USDC fleet on mainnet

- Introduced MidasArk contract for managing token supply and yield generation through Midas vaults.
- Updated deployment configurations to include MidasArk with USDC asset and associated parameters.
- Enhanced deployment artifacts and journal entries to reflect the new MidasArk integration.

halaprix committed:
feat: enhance MidasArk contract with token configuration validation

- Added checks in MidasArk to ensure payment tokens are stable and have no associated fees.
- Introduced a new error for invalid token configurations in IMidasArkErrors.
- Updated IDepositVault and IRedemptionVault interfaces to include tokensConfig function documentation.
- Modified deployment configurations for improved clarity and consistency.
